File: src/cloudsecurity_af/schemas/recon.py
"""RECON phase schemas for CloudSecurity AF.

See docs/ARCHITECTURE.md Phase 1: RECON for full specifications.
"""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# IaC Reader output
# ---------------------------------------------------------------------------


class Variable(BaseModel):
    """Terraform variable or CloudFormation parameter."""

    name: str
    type: str | None = None
    default: str | None = None
    description: str | None = None
    file_path: str | None = None


class Output(BaseModel):
    """Terraform output or CloudFormation output."""

    name: str
    value: str | None = None
    description: str | None = None
    file_path: str | None = None


class ProviderConfig(BaseModel):
    """Cloud provider block (e.g., aws, google, azurerm)."""

    name: str
    region: str | None = None
    alias: str | None = None
    version: str | None = None


class Module(BaseModel):
    """Terraform module reference."""

    name: str
    source: str
    version: str | None = None
    file_path: str | None = None


class Resource(BaseModel):
    """Individual IaC resource (Terraform resource, CloudFormation resource, K8s object)."""

    id: str = Field(description='e.g., "aws_s3_bucket.data_lake"')
    type: str = Field(description='e.g., "aws_s3_bucket"')
    name: str = Field(description='e.g., "data_lake"')
    provider: str = Field(description="aws | gcp | azure | kubernetes")
    file_path: str
    line_number: int = 0
    config: dict[str, Any] = Field(default_factory=dict, description="Raw resource configuration")
    references: list[str] = Field(
        default_factory=list,
        description="IDs of resources this depends on",
    )
    referenced_by: list[str] = Field(
        default_factory=list,
        description="IDs of resources that depend on this",
    )


class ResourceInventory(BaseModel):
    """Inventory pointer from IaC reader harness."""

    inventory_saved_path: str = Field(description="Absolute path to the generated inventory.json file")
    total_resources: int = 0
    iac_type: str = Field(
        default="terraform",
        description="terraform | cloudformation | kubernetes",
    )
    iac_version: str | None = None


# ---------------------------------------------------------------------------
# Resource Graph Builder output
# ---------------------------------------------------------------------------


class ResourceGraph(BaseModel):
    """Graph pointer from Resource Graph Builder harness."""

    graph_saved_path: str = Field(description="Absolute path to the generated graph.json file")
    total_nodes: int = 0
    total_edges: int = 0


# ---------------------------------------------------------------------------
# Drift Detection output (Tier 2+)
# ---------------------------------------------------------------------------


class ConfigDiff(BaseModel):
    """Single attribute difference between IaC and live state."""

    attribute: str
    iac_value: Any = None
    live_value: Any = None
    security_impact: str | None = None


class DriftedResource(BaseModel):
    """A resource that has drifted from its IaC declaration."""

    resource_id: str
    resource_type: str
    iac_config: dict[str, Any] = Field(default_factory=dict)
    live_config: dict[str, Any] = Field(default_factory=dict)
    diffs: list[ConfigDiff] = Field(default_factory=list)
    security_relevant: bool = False
    significance: str = Field(
        default="medium",
        description="critical | high | medium | low",
    )


class DriftReport(BaseModel):
    """Complete drift analysis between IaC and live cloud."""

    drifted_resources: list[DriftedResource] = Field(default_factory=list)
    iac_only_resources: list[str] = Field(
        default_factory=list,
        description="Declared in IaC but not deployed",
    )
    cloud_only_resources: list[str] = Field(
        default_factory=list,
        description="Deployed but not in IaC (shadow IT)",
    )


# ---------------------------------------------------------------------------
# Aggregated RECON result
# ---------------------------------------------------------------------------


class ReconResult(BaseModel):
    """Complete RECON phase output."""

    inventory: ResourceInventory = Field(default_factory=ResourceInventory)
    resource_graph: ResourceGraph = Field(default_factory=ResourceGraph)
    drift_report: DriftReport | None = Field(
        default=None,
        description="Only populated for Tier 2+ scans",
    )
    live_inventory: ResourceInventory | None = Field(
        default=None,
        description="Live cloud state (Tier 2+ only)",
    )
    iac_type: str = "terraform"
    providers_detected: list[str] = Field(default_factory=list)
    total_resources: int = 0
    total_edges: int = 0
    recon_duration_seconds: float = 0.0

File: go/internal/schemas/recon.go
package schemas

// This file ports src/cloudsecurity_af/schemas/recon.py — the RECON phase
// schemas (IaC reader output, resource-graph pointer, drift report, and the
// aggregated ReconResult).

// ---------------------------------------------------------------------------
// IaC Reader output
// ---------------------------------------------------------------------------

// Variable ports recon.py Variable: a Terraform variable or CloudFormation
// parameter.
type Variable struct {
	Name        string  `json:"name"`
	Type        *string `json:"type"`
	Default     *string `json:"default"`
	Description *string `json:"description"`
	FilePath    *string `json:"file_path"`
}

// Output ports recon.py Output: a Terraform or CloudFormation output.
type Output struct {
	Name        string  `json:"name"`
	Value       *string `json:"value"`
	Description *string `json:"description"`
	FilePath    *string `json:"file_path"`
}

// ProviderConfig ports recon.py ProviderConfig: a cloud provider block
// (e.g. aws, google, azurerm).
type ProviderConfig struct {
	Name    string  `json:"name"`
	Region  *string `json:"region"`
	Alias   *string `json:"alias"`
	Version *string `json:"version"`
}

// Module ports recon.py Module: a Terraform module reference.
type Module struct {
	Name     string  `json:"name"`
	Source   string  `json:"source"`
	Version  *string `json:"version"`
	FilePath *string `json:"file_path"`
}

// Resource ports recon.py Resource: an individual IaC resource (Terraform
// resource, CloudFormation resource, K8s object).
type Resource struct {
	// ID is e.g. "aws_s3_bucket.data_lake".
	ID string `json:"id"`
	// Type is e.g. "aws_s3_bucket".
	Type string `json:"type"`
	// Name is e.g. "data_lake".
	Name string `json:"name"`
	// Provider is one of aws | gcp | azure | kubernetes.
	Provider   string `json:"provider"`
	FilePath   string `json:"file_path"`
	LineNumber int    `json:"line_number"`
	// Config is the raw resource configuration.
	Config map[string]any `json:"config"`
	// References holds IDs of resources this depends on.
	References []string `json:"references"`
	// ReferencedBy holds IDs of resources that depend on this.
	ReferencedBy []string `json:"referenced_by"`
}

// ResourceInventory ports recon.py ResourceInventory: the inventory pointer the
// IaC reader harness returns.
//
// Python parity: InventorySavedPath is REQUIRED (no default), which is what
// makes ReconResult's default_factory unusable — see doc.go.
type ResourceInventory struct {
	// InventorySavedPath is the absolute path to the generated inventory.json.
	InventorySavedPath string `json:"inventory_saved_path"`
	TotalResources     int    `json:"total_resources"`
	// IaCType is one of terraform | cloudformation | kubernetes.
	IaCType    string  `json:"iac_type"`
	IaCVersion *string `json:"iac_version"`
}

// ---------------------------------------------------------------------------
// Resource Graph Builder output
// ---------------------------------------------------------------------------

// ResourceGraph ports recon.py ResourceGraph: the graph pointer the Resource
// Graph Builder harness returns.
//
// Python parity: GraphSavedPath is REQUIRED (no default).
type ResourceGraph struct {
	// GraphSavedPath is the absolute path to the generated graph.json.
	GraphSavedPath string `json:"graph_saved_path"`
	TotalNodes     int    `json:"total_nodes"`
	TotalEdges     int    `json:"total_edges"`
}

// ---------------------------------------------------------------------------
// Drift Detection output (Tier 2+)
// ---------------------------------------------------------------------------

// ConfigDiff ports recon.py ConfigDiff: a single attribute difference between
// IaC and live state. IaCValue/LiveValue are `Any = None` in Python.
type ConfigDiff struct {
	Attribute      string  `json:"attribute"`
	IaCValue       any     `json:"iac_value"`
	LiveValue      any     `json:"live_value"`
	SecurityImpact *string `json:"security_impact"`
}

// DriftedResource ports recon.py DriftedResource: a resource that has drifted
// from its IaC declaration.
type DriftedResource struct {
	ResourceID       string         `json:"resource_id"`
	ResourceType     string         `json:"resource_type"`
	IaCConfig        map[string]any `json:"iac_config"`
	LiveConfig       map[string]any `json:"live_config"`
	Diffs            []ConfigDiff   `json:"diffs"`
	SecurityRelevant bool           `json:"security_relevant"`
	// Significance is one of critical | high | medium | low.
	Significance string `json:"significance"`
}

// DriftReport ports recon.py DriftReport: the complete drift analysis between
// IaC and live cloud.
type DriftReport struct {
	DriftedResources []DriftedResource `json:"drifted_resources"`
	// IaCOnlyResources are declared in IaC but not deployed.
	IaCOnlyResources []string `json:"iac_only_resources"`
	// CloudOnlyResources are deployed but not in IaC (shadow IT).
	CloudOnlyResources []string `json:"cloud_only_resources"`
}

// ---------------------------------------------------------------------------
// Aggregated RECON result
// ---------------------------------------------------------------------------

// ReconResult ports recon.py ReconResult: the complete RECON phase output.
//
// Python parity: DriftReport and LiveInventory are only populated for Tier 2+
// scans. See doc.go for why Python's `ReconResult()` (no arguments) raises and
// how NewReconResult() reproduces the intended default vector instead.
type ReconResult struct {
	Inventory     ResourceInventory `json:"inventory"`
	ResourceGraph ResourceGraph     `json:"resource_graph"`
	// DriftReport is only populated for Tier 2+ scans.
	DriftReport *DriftReport `json:"drift_report"`
	// LiveInventory is the live cloud state (Tier 2+ only).
	LiveInventory        *ResourceInventory `json:"live_inventory"`
	IaCType              string             `json:"iac_type"`
	ProvidersDetected    []string           `json:"providers_detected"`
	TotalResources       int                `json:"total_resources"`
	TotalEdges           int                `json:"total_edges"`
	ReconDurationSeconds float64            `json:"recon_duration_seconds"`
}
